inst/include/Rcpp/sugar/functions/median.h

```diff
@@ -0,0 +1,296 @@
+    
+private:
+    VECTOR x;
+    
+public:
+    Median(const VECTOR& xx)
+        : x(Rcpp::clone(xx)) {}
+    
+    operator result_type() {
+        if (x.size() < 1) {
+            return Rcpp::traits::get_na<RTYPE2>();
+        }
+        
+        R_xlen_t n = x.size() / 2;
+        std::nth_element(
+            x.begin(), x.begin() + n, x.end(), 
+            median_detail::less<stored_type>);
+        
+        if (x.size() % 2) return x[n];
+        return median_detail::half(
+            x[n] + *std::max_element(
+                    x.begin(), x.begin() + n, 
+                    median_detail::less<stored_type>));
+    }
+};
+
+// specialize for character vector
+// due to string_proxy's incompatibility
+// with certain std:: algorithms;
+// need to return NA for even-length vectors
+template <bool NA, typename T, bool NA_RM>
+class Median<STRSXP, NA, T, NA_RM> {
+public:
+    typedef typename median_detail::result<STRSXP>::type result_type;
+    typedef typename Rcpp::traits::storage_type<STRSXP>::type stored_type;
+    typedef T VECTOR;
+    
+private:
+    VECTOR x;
+    
+public:
+    Median(const VECTOR& xx)
+        : x(Rcpp::clone(xx)) {}
+    
+    operator result_type() {
+        if (!(x.size() % 2)) {
+            return Rcpp::traits::get_na<STRSXP>();
+        }
+        
+        if (Rcpp::any(Rcpp::is_na(x))) {
+            return Rcpp::traits::get_na<STRSXP>();
+        }
+        
+        R_xlen_t n = x.size() / 2;
+        x.sort();
+        
+        return x[n];
+    }
+};
+
+// na.rm = TRUE
+template <bool NA, typename T>
+class Median<STRSXP, NA, T, true> {
+public:
+    typedef typename median_detail::result<STRSXP>::type result_type;
+    typedef typename Rcpp::traits::storage_type<STRSXP>::type stored_type;
+    typedef T VECTOR;
+    
+private:
+    VECTOR x;
+    
+public:
+    Median(const VECTOR& xx)
+        : x(Rcpp::na_omit(Rcpp::clone(xx))) {}
+    
+    operator result_type() {
+        if (!(x.size() % 2)) {
+            return Rcpp::traits::get_na<STRSXP>();
+        }
+        
+        R_xlen_t n = x.size() / 2;
+        x.sort();
+        
+        return x[n];
+    }
+};
+
+// NA = false
+template <typename T>
+class Median<STRSXP, false, T, true> {
+public:
+    typedef typename median_detail::result<STRSXP>::type result_type;
+    typedef typename Rcpp::traits::storage_type<STRSXP>::type stored_type;
+    typedef T VECTOR;
+    
+private:
+    VECTOR x;
+    
+public:
+    Median(const VECTOR& xx)
+        : x(Rcpp::clone(xx)) {}
+    
+    operator result_type() {
+        if (!(x.size() % 2)) {
+            return Rcpp::traits::get_na<STRSXP>();
+        }
+        
+        R_xlen_t n = x.size() / 2;
+        x.sort();
+        
+        return x[n];
+    }
+};
+
+} // sugar
+
+template <int RTYPE, bool NA, typename T>
+inline typename sugar::median_detail::result<RTYPE>::type
+median(const Rcpp::VectorBase<RTYPE, NA, T>& x, bool na_rm = false) {
+    switch (static_cast<int>(na_rm)) {
```

Comment:
How come we can't just use an `if-else` on the `na_rm` parameter? The `static_cast<int>` is kind of funky.